### ISSUE TITLE
Moving elements around unlinks connectors

### ISSUE THREAD

**Issue by Advanium** [umlet/umlet#382]
Most of the time I find myself copying elements and pasting them onto the document. The copy gets created in the top left corner of the viewport, so I often have to move it to the correct position in my diagram.

Now, it happens regularly that when I move the element(s) around and touch the connectors of other elements, those connectors get disconnected and attached to the copied elements I'm moving.
When it happens, I have to stop the move, and undo a couple of times to restore the state of my diagram.

I'm sure it also happened to me before when moving elements from one side of the diagramm to the other side.

**Comment by afdia** [comment]
Hi

Umlet doesnt track connections between elements and relations. It simply moves relation ends if they are near the border of an element.

Perhaps we should communicate this feature better, but you can hold SHIFT while dragging an element to temporarily disable sticking. That way you simply move if you want relations to stick and you SHIFT+move if you want to move the element without any relation ends.

**Comment by Advanium** [comment]
I think you misunderstood. Sorry for not expressing my issue better. I made a video, see https://youtu.be/3L7NKK60oWo

**Comment by afdia** [comment]
Thanks for the video, I see the problem now.

Are you allowed to share the specific diagram from the video (e.g. upload the uxf xml content as a Gist) or perhaps you can create a minimal diagram with a short step by step list to reproduce the issue

I tried it myself but couldn't reproduce the "losing the connection" behavior.
Perhaps it has to do with some specific size of some elements or other properties of this diagram.

**Comment by Advanium** [comment]
I tried it a couple of times and it seems to be reproducable, f.e. with the specific node from the video.
Here's the gist containing the test file.
https://gist.github.com/Advanium/e2d0f662e29a0e3220054ece0d6b0089
Thanks for looking into this.

**Comment by afdia** [comment]
Thanks I can reproduce the issue now.
I've also reduced the diagram even more:
https://gist.github.com/afdia/cbb98f4d715adf48d632838425a1235b

Steps to reproduce the problem:
1. drag upper usecase down vertically
2. after the relation shrinks to 10px (1 gridsize) it loses the connection
3. Afterwards it's possible to move the usecase around without the relation as long as the usecase border doesn't touch the relation anywhere (this would "reattach" it at this location which is wrong but only a sideeffect of how relations work)
4. If moving very fast, it's also possible to "escape" without reattaching (perhaps because multiple gridsizes are moved which effectively skips the short-time-overlap of the usecase edge)

Both 3 and 4 should never happen if the connection is not lost, so this is the root cause of the issue

**Comment by afdia** [comment]
I found the problem.
When creating the new relations, I added a safety check which basically says "if a relation would disappear completely (only 2 points which are overlapping) undo the last change"

The problem is that everytime you accidently overlap both ends of a simple relation while dragging, the last move will be undone and you will "lose" the connection.

At first I tried to fix it by doing this check only after the whole continuous move is finished (i.e. the user releases the mouse button), but currently there is no trigger when this happens and it would still cause problems with the undo-function (because this safety check is not part of the drag-command)

After all I removed this safety check completely which fixes the issue.
Of course you have to avoid moving both endpoints of a relation at the same space, because such a relation wouldn't be visible anymore (but still clickable to be removed), but this is the same behavior as the with the old relations.

Thanks again for the report, I always had the feeling that sometimes connections are lost, but it's hard to locate the problem in complex gui code such as the relation-sticking-behavior.
